url80587fc7d5_dennervidal_CloudSim_DVFS_tgz-pJ8-Examples_Dvfs_example_simpleJ8/src/org/cloudbus/cloudsim/core/CloudSim.java

```diff
@@ -0,0 +1,928 @@
+		traceFlag = _traceFlag;
+
+		// Set the current Wall clock time as the starting time of
+		// simulation
+		if (_calendar == null) {
+			calendar = Calendar.getInstance();
+		} else {
+			calendar = _calendar;
+		}
+
+		// creates a CloudSimShutdown object
+		CloudSimShutdown shutdown = new CloudSimShutdown("CloudSimShutdown", numUser);
+		shutdownId = shutdown.getId();
+	}
+
+	/**
+	 * Initialises CloudSim parameters. This method should be called before creating any entities.
+	 * <p>
+	 * Inside this method, it will create the following CloudSim entities:
+	 * <ul>
+	 * <li>CloudInformationService.
+	 * <li>CloudSimShutdown
+	 * </ul>
+	 * <p>
+	 * 
+	 * @param numUser the number of User Entities created. This parameters indicates that
+	 *            {@link gridsim.CloudSimShutdown} first waits for all user entities's
+	 *            END_OF_SIMULATION signal before issuing terminate signal to other entities
+	 * @param cal starting time for this simulation. If it is <tt>null</tt>, then the time will be
+	 *            taken from <tt>Calendar.getInstance()</tt>
+	 * @param traceFlag <tt>true</tt> if CloudSim trace need to be written
+	 * @see gridsim.CloudSimShutdown
+	 * @see CloudInformationService.CloudInformationService
+	 * @pre numUser >= 0
+	 * @post $none
+	 */
+	public static void init(int numUser, Calendar cal, boolean traceFlag) {
+		try {
+			initCommonVariable(cal, traceFlag, numUser);
+
+			// create a GIS object
+			cis = new CloudInformationService("CloudInformationService");
+
+			// set all the above entity IDs
+			cisId = cis.getId();
+		} catch (IllegalArgumentException s) {
+			Log.printLine("CloudSim.init(): The simulation has been terminated due to an unexpected error");
+			Log.printLine(s.getMessage());
+		} catch (Exception e) {
+			Log.printLine("CloudSim.init(): The simulation has been terminated due to an unexpected error");
+			Log.printLine(e.getMessage());
+		}
+	}
+
+	/**
+	 * Starts the execution of CloudSim simulation. It waits for complete execution of all entities,
+	 * i.e. until all entities threads reach non-RUNNABLE state or there are no more events in the
+	 * future event queue.
+	 * <p>
+	 * <b>Note</b>: This method should be called after all the entities have been setup and added.
+	 * 
+	 * @return the double
+	 * @throws NullPointerException This happens when creating this entity before initialising
+	 *             CloudSim package or this entity name is <tt>null</tt> or empty.
+	 * @see gridsim.CloudSim#init(int, Calendar, boolean)
+	 * @pre $none
+	 * @post $none
+	 */
+	public static double startSimulation() throws NullPointerException {
+		Log.printLine("Starting CloudSim version " + CLOUDSIM_VERSION_STRING);
+		try {
+			double clock = run();
+
+			// reset all static variables
+			cisId = -1;
+			shutdownId = -1;
+			cis = null;
+			calendar = null;
+			traceFlag = false;
+
+			return clock;
+		} catch (IllegalArgumentException e) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: While wrapping the caught exception into a custom one, information about the caught exception is being lost, including information about the stack trace of the exception.

**Fix**: If the caught exception object does not contain sensitive information, consider passing it as the &#34;rootCause&#34; or inner exception parameter to the constructor of the new exception before throwing the new exception. (Note that not all exception constructors support inner exceptions. Use a wrapper exception that supports inner exceptions.) 
[Learn more](https://www.ibm.com/support/pages/best-practice-catching-and-re-throwing-java-exceptions)